src/scheduler/job_info.c

```diff
@@ -2709,64 +2709,71 @@ preempt_job_set_filter(resource_resv *job, void *arg)
 
 /**
  * @brief
- *  	get_preemption_order - deduce the preemption ordering to be used for a job
+ *  get_job_req_used_time - get a running job's req and used time for preemption
  *
- * @param[in]	pjob	-	the job to preempt
- * @param[in]	sinfo	-	Pointer to server info structure.
- *
- * @return	: struct preempt_ordering.  array containing preemption order
+ * @param[in]	pjob - the job in question
+ * @param[out]	rtime - return pointer to the requested time
+ * @param[out]	utime - return pointer to the used time
  *
+ * @return	int
+ * @retval	0 for success
+ * @retval	1 for error
  */
-struct preempt_ordering *get_preemption_order(resource_resv *pjob,
-	server_info *sinfo)
+static int
+get_job_req_used_time(resource_resv *pjob, double *rtime, double *utime)
 {
-	/* the order to preempt jobs in */
-	struct preempt_ordering *po = &conf.preempt_order[0];
-	int i;
+	resource_req *req; /* the jobs requested soft_walltime/walltime/cput */
+	resource_req *used; /* the amount of the walltime/cput used */
 
-	if (pjob == NULL || pjob->job == NULL)
-		return 0;
+	if (pjob == NULL || pjob->job == NULL || !pjob->job->is_running
+			|| rtime == NULL || utime == NULL)
+		return 1;
 
-	/* continue validity checks */
-	if (!pjob->job->is_running || pjob->ninfo_arr == NULL)
-		return 0;
+	req = find_resource_req(pjob->resreq, getallres(RES_SOFT_WALLTIME));
 
-	/* check if we have more then one range... no need to choose if not */
-	if (conf.preempt_order[1].high_range != 0) {
-		resource_req *req;		/* the jobs requested soft_walltime/walltime/cput */
-		resource_req *used;		/* the amount of the walltime/cput used */
+	if (req == NULL)
+		req = find_resource_req(pjob->resreq, getallres(RES_WALLTIME));
 
-		req = find_resource_req(pjob->resreq, getallres(RES_SOFT_WALLTIME));
+	if (req == NULL) {
+		req = find_resource_req(pjob->resreq, getallres(RES_CPUT));
+		used = find_resource_req(pjob->job->resused, getallres(RES_CPUT));
+	} else
+		used = find_resource_req(pjob->job->resused, getallres(RES_WALLTIME));
 
-		if(req == NULL)
-			req = find_resource_req(pjob->resreq, getallres(RES_WALLTIME));
+	if (req != NULL && used != NULL) {
+		*rtime = (double) req->amount;
+		*utime = (double) used->amount;
+	}
```

Comment:
Shouldn't there be an else case here to return 1 if we find even one (or both) to be NULL?

Reply:
no, when either of them is NULL, we are supposed to return the first preempt order, see get_preempt_order, it returns the first preempt order if req or used are less than 0. So, actually, we should set rtime and utime to -1 in the else case.